sql/core/src/test/scala/org/apache/spark/sql/execution/exchange/EnsureRequirementsSuite.scala

```diff
@@ -135,4 +137,489 @@ class EnsureRequirementsSuite extends SharedSparkSession {
+        case other => fail(other.toString)
+      }
+
+      // ShuffleExchange(7) <-> HashPartitioning(6)
+      // Should shuffle LHS and convert to HashPartitioning(6) <-> HashPartitioning(6)
+      plan1 = ShuffleExchangeExec(
+        outputPartitioning = HashPartitioning(exprA :: exprB :: Nil, 7),
+        child = DummySparkPlan())
+      plan2 = DummySparkPlan(
+        outputPartitioning = HashPartitioning(exprC :: exprD :: Nil, 6))
+      smjExec = SortMergeJoinExec(
+        exprA :: exprB :: Nil, exprC :: exprD :: Nil, Inner, None, plan1, plan2)
+      EnsureRequirements.apply(smjExec) match {
+        case SortMergeJoinExec(_, _, _, _,
+        SortExec(_, _, ShuffleExchangeExec(left: HashPartitioning, _, _), _),
+        SortExec(_, _, DummySparkPlan(_, _, right: HashPartitioning, _, _), _), _) =>
+          assert(left.expressions === Seq(exprA, exprB))
+          assert(right.expressions === Seq(exprC, exprD))
+          assert(left.numPartitions == 6)
+          assert(right.numPartitions == 6)
+        case other => fail(other.toString)
+      }
+
+      // ShuffleExchange(7) <-> ShuffleExchange(6)
+      // Should consider `spark.sql.shuffle.partitions` and shuffle both sides, and
+      // convert to HashPartitioning(10) <-> HashPartitioning(10)
+      plan1 = ShuffleExchangeExec(
+        outputPartitioning = HashPartitioning(exprA :: Nil, 7),
+        child = DummySparkPlan())
+      plan2 = ShuffleExchangeExec(
+        outputPartitioning = HashPartitioning(exprC :: Nil, 6),
+        child = DummySparkPlan())
+      smjExec = SortMergeJoinExec(
+        exprA :: exprB :: Nil, exprC :: exprD :: Nil, Inner, None, plan1, plan2)
+      EnsureRequirements.apply(smjExec) match {
+        case SortMergeJoinExec(_, _, _, _,
+        SortExec(_, _, ShuffleExchangeExec(left: HashPartitioning, _, _), _),
+        SortExec(_, _, ShuffleExchangeExec(right: HashPartitioning, _, _), _), _) =>
+          assert(left.expressions === Seq(exprA, exprB))
+          assert(right.expressions === Seq(exprC, exprD))
+          assert(left.numPartitions == conf.numShufflePartitions)
+          assert(right.numPartitions == conf.numShufflePartitions)
+        case other => fail(other.toString)
+      }
+    }
+  }
+
+  test("Respect spark.sql.shuffle.partitions with AQE") {
+    withSQLConf(SQLConf.SHUFFLE_PARTITIONS.key -> 8.toString,
+      SQLConf.COALESCE_PARTITIONS_INITIAL_PARTITION_NUM.key -> 10.toString) {
+      Seq(false).foreach { enable =>
+        withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> s"$enable") {
+          val plan1 = DummySparkPlan(
+            outputPartitioning = HashPartitioning(exprA :: exprB :: Nil, 9))
+          val plan2 = DummySparkPlan(
+            outputPartitioning = UnknownPartitioning(8))
+          val smjExec = SortMergeJoinExec(
+            exprA :: exprB :: Nil, exprC :: exprD :: Nil, Inner, None, plan1, plan2)
+          EnsureRequirements.apply(smjExec) match {
```

Comment:
Does AQE play a role here? This is UT and the code path being tested is not affected by AQE at all.